stream-legal-entity/legal-entity-core/src/main/java/com/backbase/stream/LegalEntitySaga.java

```diff
@@ -583,19 +589,25 @@ private User setupAdminInternalId(LegalEntity legalEntity, JobProfileUser jobPro
                     streamTask.info(SERVICE_AGREEMENT, SETUP_SERVICE_AGREEMENT, EXISTS, serviceAgreement.getExternalId(), serviceAgreement.getInternalId(), "Existing Service Agreement: %s found for Legal Entity: %s", serviceAgreement.getExternalId(), legalEntity.getExternalId());
                     return Mono.just(streamTask);
                 });
-            ServiceAgreement newServiceAgreement = createMasterServiceAgreement(legalEntity, legalEntity.getAdministrators());
 
-            Mono<LegalEntityTask> createServiceAgreement = accessGroupService.createServiceAgreement(streamTask, newServiceAgreement)
-                .onErrorMap(AccessGroupException.class, accessGroupException -> {
-                    streamTask.error(SERVICE_AGREEMENT, SETUP_SERVICE_AGREEMENT, FAILED, newServiceAgreement.getExternalId(), null, accessGroupException, accessGroupException.getMessage(), accessGroupException.getHttpResponse());
-                    return new StreamTaskException(streamTask, accessGroupException);
-                })
-                .flatMap(serviceAgreement -> {
-                    streamTask.getData().setMasterServiceAgreement(serviceAgreement);
-                    streamTask.info(SERVICE_AGREEMENT, SETUP_SERVICE_AGREEMENT, CREATED, serviceAgreement.getExternalId(), serviceAgreement.getInternalId(), "Created new Service Agreement: %s with Administrators: %s for Legal Entity: %s", serviceAgreement.getExternalId(), legalEntity.getAdministrators().stream().map(User::getExternalId).collect(Collectors.joining(", ")), legalEntity.getExternalId());
-                    return Mono.just(streamTask);
-                });
-            return existingServiceAgreement.switchIfEmpty(createServiceAgreement);
+                // Master Service Agreement can be created only if activateSingleServiceAgreement property is missing or it has the value: true
+                if (streamTask.getLegalEntity() != null &&
+                    (streamTask.getLegalEntity().getActivateSingleServiceAgreement() == null || streamTask.getLegalEntity().getActivateSingleServiceAgreement())) {
+                    ServiceAgreement newServiceAgreement = createMasterServiceAgreement(legalEntity, legalEntity.getAdministrators());
+                    Mono<LegalEntityTask> createServiceAgreement = accessGroupService.createServiceAgreement(streamTask, newServiceAgreement)
+                        .onErrorMap(AccessGroupException.class, accessGroupException -> {
+                            streamTask.error(SERVICE_AGREEMENT, SETUP_SERVICE_AGREEMENT, FAILED, newServiceAgreement.getExternalId(), null, accessGroupException, accessGroupException.getMessage(), accessGroupException.getHttpResponse());
+                            return new StreamTaskException(streamTask, accessGroupException);
+                        })
+                        .flatMap(serviceAgreement -> {
```

Comment:
Why is this a flatmap when you are not making a call that requires IO?

Reply:
@bartveenstra that was existing logic. I haven't changed that. All I did was, add an if condition around it, that master service agreement should only created when ActivateSingleServiceAgreement is set to true or null.